x/evm/keeper/state_transition.go

```diff
@@ -560,36 +581,29 @@ func (k *Keeper) prepareTxForSgx(ctx sdk.Context, msg core.Message, cfg *EVMConf
 			DebugTrace:      cfg.DebugTrace,
 			NoBaseFee:       cfg.FeeMarketParams.NoBaseFee,
 			EvmDenom:        cfg.Params.EvmDenom,
-			Overrides:       ToGOB64(cfg.Overrides),
+			Overrides:       overrides,
 		},
 	}
 
-	return sgxRPCClient.PrepareTx(args, &PrepareTxReply{})
-}
-
-func (k *Keeper) runRPCServer(ctx sdk.Context, msg core.Message, cfg *EVMConfig) error {
-	defer func() {
-		if r := recover(); r != nil {
-			ctx.Logger().Debug("recovered from panic", "error", r)
-		}
-	}()
-
-	// TODO Think about whether the RPC server should be persistent or ephemeral
-	//  - If it's persistent, we need to handle the lifecycle of the RPC server
-	//  - If it's ephemeral, we need to create a new RPC server for each message
-	//  The current implementation is ephemeral.
-	srv := &EthmRpcServer{k: k, ctx: ctx, msg: msg, evmCfg: cfg}
-	rpc.Register(srv)
-	rpc.HandleHTTP()
-
-	// TODO handle port customization
-	l, err := net.Listen("tcp", ":9093")
+	resp, err := sgxGrpcClient.PrepareTx(args)
 	if err != nil {
-		// TODO handle error
-		panic(err)
+		// panic cosmos if sgx isn't available.
+		if k.IsSgxDownError(err) {
+			panic("sgx rpc server is down")
+		}
+
+		return 0, err
 	}
-	// TODO Handle shutdown
-	go http.Serve(l, nil)
 
-	return nil
+	// Store handler id
+	handlerId := resp.HandlerId
+	// Snapshot the ctx
+	k.sdkCtxs[handlerId] = &ctx
```

Comment:
Do we need a mutex here?

Reply:
@amaury1093 not sure exactly. I guess it was working without problem last time we managed to run with rpc.